modules/core/04-channel/v2/keeper/packet.go

```diff
@@ -173,6 +162,62 @@ func (k Keeper) recvPacket(
 	return nil
 }
 
+func (k Keeper) acknowledgePacket(ctx context.Context, packet channeltypesv2.Packet, acknowledgement channeltypesv2.Acknowledgement, proof []byte, proofHeight exported.Height) error {
+	// Lookup counterparty associated with our channel and ensure
+	// that the packet was indeed sent by our counterparty.
+	counterparty, ok := k.GetCounterparty(ctx, packet.SourceId)
+	if !ok {
+		return errorsmod.Wrap(types.ErrCounterpartyNotFound, packet.SourceId)
+	}
+
+	if counterparty.ClientId != packet.DestinationId {
+		return channeltypes.ErrInvalidChannelIdentifier
+	}
+	clientID := counterparty.ClientId
+
+	commitment := k.GetPacketCommitment(ctx, packet.SourceId, packet.Sequence)
+	if len(commitment) == 0 {
+		// TODO: emit events
+		// k.EmitAcknowledgePacketEvent(ctx, packet, nil)
+
+		// This error indicates that the acknowledgement has already been relayed
+		// or there is a misconfigured relayer attempting to prove an acknowledgement
+		// for a packet never sent. Core IBC will treat this error as a no-op in order to
+		// prevent an entire relay transaction from failing and consuming unnecessary fees.
+		return channeltypes.ErrNoOpMsg
+	}
+
+	packetCommitment := channeltypesv2.CommitPacket(packet)
+
+	// verify we sent the packet and haven't cleared it out yet
+	if !bytes.Equal(commitment, packetCommitment) {
+		return errorsmod.Wrapf(channeltypes.ErrInvalidPacket, "commitment bytes are not equal: got (%v), expected (%v)", packetCommitment, commitment)
+	}
+
+	path := hostv2.PacketAcknowledgementKey(packet.DestinationId, sdk.Uint64ToBigEndian(packet.Sequence))
+	merklePath := types.BuildMerklePath(counterparty.MerklePathPrefix, path)
```

Comment:
`BuildMerklePath`, used here and other funcs in this file is imported from the packet-server package.

We should move this to the channel/v2 types pkg afaik. We should not depend on anything in packet-server (which will ultimately be removed).

I'd also like some insight into the implementation of `BuildMerklePath`.
Some questions I have are:
Why are we not using the `MerklePrefix` type for the "ibc" prefix used by counterparties - instead we're using `MerklePath`. 
Secondly, implementation looks complex and kinda funky, haven't stepped through to see exactly why that is, was hoping someone could explain it for me.

Reply:
re dependencies: I think we can do a pass at the end and make sure there is no dependency on v1 anything within the channel/v2 package.

Reply:
cc. @AdityaSripal @DimitrisJim on this func